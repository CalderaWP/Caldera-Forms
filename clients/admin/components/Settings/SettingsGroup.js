import React from 'react';
import PropTypes from 'prop-types';
import {object, pick} from "dot-object";
import {RenderGroup} from '@caldera-labs/components';
import {Button} from '@wordpress/components';
import classNames from 'classnames'
export class SettingsGroup extends React.PureComponent{

	constructor(props) {
		super(props);
		this.state = {
			[props.settingsKey]: props.settings
		};
		this.onSettingsChange = this.onSettingsChange.bind(this);
		this.getConfigFields = this.getConfigFields.bind(this);
		this.onSettingsSave = this.onSettingsSave.bind(this);
		this.wrapperClass = this.wrapperClass.bind(this);

	}

	/**
	 * Save the settings
	 */
	onSettingsSave() {
		this.props.onSettingsSave(this.state[this.props.settingsKey]);
	}

	/**
	 * Update internal state when settings change
	 *
	 * @param {Object} update
	 */
	onSettingsChange(update) {
		this.setState(
			{
				[this.props.settingsKey]:update
		});
	}

	/**
	 * Prepare config fields
	 *
	 * @return {Array}
	 */
	getConfigFields() {
		let currentConfigFields = this.props.configFields;
		currentConfigFields.forEach(configField => {
			const {path} = configField;
			if( undefined !== path ){
				configField.value = pick(
					path,
					this.state
				);
				configField.onValueChange = (newValue) => {
					const update = {
						...this.state,
						[path]:newValue
					};
					this.onSettingsChange(object(update));
				};
			}





		});
		return currentConfigFields;
	};

	/**
	 * Get the class for the outer element
	 * @return {String}
	 */
	wrapperClass(){
		return this.props.wrapperClass;
	}


	render(){
		if( Array.isArray( this.props.configFields ) ) {
			return(
				<div
					className={classNames(
						this.wrapperClass(),
						this.props.classNames
					)}
				>
					<RenderGroup
						configFields={this.getConfigFields()}
					/>
					<Button
						onClick={this.props.onSettingsSave}
					>
						{this.props.saveButtonText}
					</Button>
				</div>
			)
		}
		return <div>!</div>

	}

}

/**
 * Prop types for the SettingsGroup component
 * @type {{}}
 */
SettingsGroup.propTypes = {
	classNames: PropTypes.string,
	settings: PropTypes.object.isRequired,
	settingsKey: PropTypes.string.isRequired,
	onSettingsSave: PropTypes.func.isRequired,
	configFields: PropTypes.array,
	saveButtonText: PropTypes.string,
};

SettingsGroup.defaultProps= {
	saveButtonText: 'Save'
}