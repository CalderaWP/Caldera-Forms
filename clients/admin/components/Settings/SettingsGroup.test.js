import renderer from 'react-test-renderer';
import React from 'react';
import {shallow} from 'enzyme';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import {SettingsGroup} from "./SettingsGroup";

Enzyme.configure({adapter: new Adapter()});
const handler = () => {
};

describe('Settings component', () => {
	const mockSettings = {
		roy: 'Sivan'
	};

	const configFields = [
		{
			id: 'roy',
			onValueChange: handler,
			value: 3,
			type: 'number',
			path: 'mockStuff.roy'
		}
	];

	it('Renders with minimal props', () => {
		const componet = renderer.create(
			<SettingsGroup
				onSettingsSave={handler}
				settings={mockSettings}
				settingsKey={'mockStuff'}
				configFields={configFields}
			/>
		);
		expect(componet.toJSON()).toMatchSnapshot();
	});
	it('Sets settings in state', () => {
		const componet = shallow(
			<SettingsGroup
				onSettingsSave={handler}
				settings={mockSettings}
				settingsKey={'mockStuff'}
				configFields={configFields}
			/>
		);
		expect(componet.state('mockStuff')).toEqual(mockSettings);
	});

	it('Passes updates to state to save handle', () => {
		const updatedSettings = {
			roy: true
		};

		let updates = {};
		const component = shallow(
			<SettingsGroup
				onSettingsSave={(values) => {
					updates = values;
				}}
				settings={mockSettings}
				settingsKey={'mockStuff'}
				configFields={[
					{
						id: 'roy',
						onValueChange: handler,
						value: 3,
						type: 'number'
					}
				]}

			/>
		);

		component.setState({mockStuff: updatedSettings});
		component.instance().onSettingsSave();
		expect(updates.roy).toEqual(true);
	});

	it('Merges settings on update', () => {
		const updatedSettings = {
			roy: true
		};

		const component = shallow(
			<SettingsGroup
				onSettingsSave={handler}
				settings={mockSettings}
				settingsKey={'mockStuff'}
				configFields={[
					{
						id: 'roy',
						onValueChange: handler,
						value: 3,
						type: 'number'
					}
				]}

			/>
		);

		component.instance().onSettingsChange(updatedSettings);
		expect(component.state('mockStuff')).toEqual(updatedSettings);
	});

});