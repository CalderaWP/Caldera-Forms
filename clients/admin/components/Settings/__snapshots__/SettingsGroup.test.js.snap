// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Settings component Renders with minimal props 1`] = `
<div
  className=""
>
  <div
    className="caldera-config-field-setup"
  >
    <div
      className="caldera-config-group"
    >
      <div
        className="caldera-config-field"
      >
        <label
          htmlFor="roy"
        />
        <input
          aria-describedby={null}
          className="field-config field-config"
          disabled={false}
          id="roy"
          onBlur={undefined}
          onChange={[Function]}
          onFocus={undefined}
          required={false}
          type="number"
          value="Sivan"
        />
      </div>
    </div>
  </div>
  <button
    autoFocus={undefined}
    className="components-button"
    disabled={undefined}
    onClick={[Function]}
    type="button"
  >
    Save
  </button>
</div>
`;
